sdks/python/apache_beam/runners/portability/fn_api_runner/watermark_manager.py

```diff
@@ -0,0 +1,171 @@
+from typing import Union
+
+from apache_beam.portability.api import beam_runner_api_pb2
+from apache_beam.runners.portability.fn_api_runner import translations
+from apache_beam.runners.portability.fn_api_runner.translations import split_buffer_id
+from apache_beam.runners.worker import bundle_processor
+from apache_beam.utils import proto_utils
+from apache_beam.utils import timestamp
+
+
+class WatermarkManager(object):
+  """Manages the watermarks of a pipeline's stages.
+    It works by constructing an internal graph representation of the pipeline,
+    and keeping track of dependencies."""
+  class PCollectionNode(object):
+    def __init__(self, name):
+      self.name = name
+      self._watermark = timestamp.MIN_TIMESTAMP
+      self.producers: Set[WatermarkManager.StageNode] = set()
+
+    def __str__(self):
+      return 'PCollectionNode<producers=%s>' % list(self.producers)
+
+    def set_watermark(self, wm: timestamp.Timestamp):
+      self._watermark = min(self.upstream_watermark(), wm)
+
+    def upstream_watermark(self):
+      if self.producers:
+        return min(p.input_watermark() for p in self.producers)
+      else:
+        return timestamp.MAX_TIMESTAMP
+
+    def watermark(self):
+      return self._watermark
+
+  class StageNode(object):
+    def __init__(self, name):
+      # We keep separate inputs and side inputs because side inputs
+      # should hold back a stage's input watermark, to hold back execution
+      # for that stage; but they should not be considered when calculating
+      # the output watermark of the stage, because only the main input
+      # can actually advance that watermark.
+      self.name = name
+      self.inputs: Set[WatermarkManager.PCollectionNode] = set()
+      self.side_inputs: Set[WatermarkManager.PCollectionNode] = set()
+      self.outputs: Set[WatermarkManager.PCollectionNode] = set()
+
+    def __str__(self):
+      return 'StageNode<inputs=%s,side_inputs=%s' % (
+          [i.name for i in self.inputs], [i.name for i in self.side_inputs])
+
+    def output_watermark(self):
+      if not self.outputs:
+        return self.input_watermark()
+      else:
+        return min(o.watermark() for o in self.outputs)
+
+    def input_watermark(self):
+      if not self.inputs:
+        return timestamp.MAX_TIMESTAMP
+      w = min(i.upstream_watermark() for i in self.inputs)
+
+      if self.side_inputs:
+        w = min(w, min(i.upstream_watermark() for i in self.side_inputs))
+      return w
+
+  def __init__(self, stages):
+    # type: (List[translations.Stage]) -> None
+    self._pcollections_by_name: Dict[Union[str, translations.TimerFamilyId],
+                                     WatermarkManager.PCollectionNode] = {}
+    self._stages_by_name: Dict[str, WatermarkManager.StageNode] = {}
+
+    for s in stages:
+      stage_name = s.name
+      stage_node = WatermarkManager.StageNode(stage_name)
+      self._stages_by_name[stage_name] = stage_node
+
+      def add_pcollection(
```

Comment:
Should this function declared outside of the loop?

Reply:
done